README.md
# Flock Postgres

Flock Postgres is a Flock plugin for Postgres.

## Install

```
npm install launchfort/flock-pg
```

## Usage

```js
// .flockrc.js
const { DefaultMigrator, NodeModuleMigrationProvider } = require('@launchfort/flock')
const { DataAccessProvider, TemplateProvider } = require('@launchfort/flock-pg')

const migrationDir = 'migrations'
const migrationTableName = 'migration'
const dap = new DataAccessProvider({ migrationTableName })
const mp = new NodeModuleMigrationProvider({ migrationDir })

exports.migrator = new DefaultMigrator(mp, dap)
exports.migrationDir = migrationDir
exports.templateProvider = new TemplateProvider()

```

## Migrations

When writing migrations that use `flock-pg` the `QueryInterface#query`
method can be used in a similar fashion as the query method from the [pg](https://node-postgres.com/features/queries)
module.

Example:
```js
exports.up = queryInterface => {
  const sql = 'SELECT * FROM user WHERE age = $1'
  const values = [ 1 ]
  return queryInterface.query({ text: sql, values })
}
```

The `QueryInterface#query` method accepts anything that the `pg` query API accepts or a query object
with the following shape:

```ts
{
  text: string,
  values?: any[],
  name?: string
}
```

## API

Flock pg exports implementations of Flock's `DataAccessProvider` and `TemplateProvider`
as `DataAccessProvider` and `TemplateProvider` classes.

The `DataAccessProvider` class will connect to your Postgres DB by reading
the connection string from the `DATABASE_URL` environment variable. Optionally you
can override the behaviour by passing in the `connectionString` option to the
constructor.

```js
class DataAccessProvider implements Flock.DataAccessProvider {
  constructor ({
    migrationTableName = 'migration',
    acquireLock = true,
    connectionString = process.env.DATABASE_URL } = {})
}
```

Additionally, by default the `DataAccessProvider` will attempt to acquire an
application lock immediately after connecting to the database. This behaviour
can be overridden by setting the `acquireLock` option to `false`. Acquiring a
lock helps to prevent concurrent migrations from occuring.

See: https://devcenter.heroku.com/articles/release-phase


CHANGELOG.md
# 1.0.0

**Major**

- Reach stability
- Update scope to @launchfort from @gradealabs

**Patch**

- Update dependencies


# 0.0.3

**Patch**

- Resolve issue with rolling back without first migrating (i.e. the migration
  table did not yet exist).

# 0.0.2

**Patch**

- Fix author email address in `package.json`.
- Remove `tsconfig-paths` from `package.json`.
- Add `.eslintignore` to ignore the `lib` folder.


lib/index.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Assert = require("assert");
const Path = require("path");
const index_1 = require("./index");
describe('flock-pg', function () {
    const dap = new index_1.DataAccessProvider();
    let da = null;
    let qi = null;
    beforeEach(function () {
        return __awaiter(this, void 0, void 0, function* () {
            da = yield dap.provide();
            qi = da['qi']; // the QueryInterface
        });
    });
    afterEach(function () {
        return __awaiter(this, void 0, void 0, function* () {
            if (!qi)
                return;
            qi.query({
                text: `DROP TABLE IF EXISTS ${da.migrationTableName}`
            });
            yield da.close();
        });
    });
    describe('TemplateProvider', function () {
        it('should provide a template file name when given a migration type that matches a template name', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const tp = new index_1.TemplateProvider();
                let fileName = yield tp.provideFileName('create-table');
                Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/create-table.ejs'));
                fileName = yield tp.provideFileName('alter-table');
                Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/alter-table.ejs'));
                fileName = yield tp.provideFileName('other');
                Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/other.ejs'));
            });
        });
        it('should reject when given a migration type that does not match a template name', function () {
            return __awaiter(this, void 0, void 0, function* () {
                const tp = new index_1.TemplateProvider();
                try {
                    yield tp.provideFileName('nope');
                }
                catch (error) {
                    Assert.strictEqual(error.code, 'UNSUPPORTED_MIGRATION_TYPE');
                }
            });
        });
    });
    describe('DataAccessProvider#provide', function () {
        it('should connect to the DB and acquire application lock', function () {
            return __awaiter(this, void 0, void 0, function* () {
                Assert.strictEqual(dap.migrationTableName, 'migration');
            });
        });
    });
    describe('PgDataAccess', function () {
        describe('#getMigratedMigrations', function () {
            it('should retrieve migrated migrations', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    yield qi.query({
                        text: `CREATE TABLE IF NOT EXISTS "${da.migrationTableName}" (
              id varchar(512),
              created_at timestamp DEFAULT current_timestamp,
              PRIMARY KEY(id)
            )`
                    });
                    yield qi.query({
                        text: `INSERT INTO "${da.migrationTableName}" (id) VALUES($1)`,
                        values: ['one']
                    });
                    const migrated = yield da.getMigratedMigrations();
                    Assert.deepStrictEqual(migrated.map(x => x.id), ['one']);
                    Assert.ok(migrated[0].migratedAt instanceof Date);
                });
            });
        });
        describe('#migrate', function () {
            it('should migrate a migration', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    yield da.migrate('two', qi => {
                        /* do nothing */
                        return Promise.resolve();
                    });
                    const migrated = yield da.getMigratedMigrations();
                    Assert.deepStrictEqual(migrated.map(x => x.id), ['two']);
                    Assert.ok(migrated[0].migratedAt instanceof Date);
                });
            });
        });
        describe('#rollback', function () {
            it('should rollback a migration', function () {
                return __awaiter(this, void 0, void 0, function* () {
                    yield da.migrate('two', qi => {
                        /* do nothing */
                        return Promise.resolve();
                    });
                    yield da.rollback('two', qi => {
                        /* do nothing */
                        return Promise.resolve();
                    });
                    const migrated = yield da.getMigratedMigrations();
                    Assert.strictEqual(migrated.length, 0);
                });
            });
        });
    });
});


src/index.test.ts
import * as Assert from 'assert'
import * as Path from 'path'
import { DataAccessProvider, PgDataAccess, PgQueryInterface, TemplateProvider } from './index'

describe('flock-pg', function () {
  const dap = new DataAccessProvider()
  let da: PgDataAccess = null
  let qi: PgQueryInterface = null

  beforeEach(async function () {
    da = await dap.provide()
    qi = da['qi'] // the QueryInterface
  })

  afterEach(async function () {
    if (!qi) return
    qi.query({
      text: `DROP TABLE IF EXISTS ${da.migrationTableName}`
    })
    await da.close()
  })

  describe('TemplateProvider', function () {
    it('should provide a template file name when given a migration type that matches a template name', async function () {
      const tp = new TemplateProvider()
      let fileName = await tp.provideFileName('create-table')
      Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/create-table.ejs'))
      fileName = await tp.provideFileName('alter-table')
      Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/alter-table.ejs'))
      fileName = await tp.provideFileName('other')
      Assert.strictEqual(fileName, Path.resolve(__dirname, './templates/other.ejs'))
    })

    it('should reject when given a migration type that does not match a template name', async function () {
      const tp = new TemplateProvider()
      try {
        await tp.provideFileName('nope')
      } catch (error) {
        Assert.strictEqual(error.code, 'UNSUPPORTED_MIGRATION_TYPE')
      }
    })
  })

  describe('DataAccessProvider#provide', function () {
    it('should connect to the DB and acquire application lock', async function () {
      Assert.strictEqual(dap.migrationTableName, 'migration')
    })
  })

  describe('PgDataAccess', function () {
    describe('#getMigratedMigrations', function () {
      it('should retrieve migrated migrations', async function () {
        await qi.query({
          text:
            `CREATE TABLE IF NOT EXISTS "${da.migrationTableName}" (
              id varchar(512),
              created_at timestamp DEFAULT current_timestamp,
              PRIMARY KEY(id)
            )`
        })
        await qi.query({
          text: `INSERT INTO "${da.migrationTableName}" (id) VALUES($1)`,
          values: [ 'one' ]
        })
        const migrated = await da.getMigratedMigrations()
        Assert.deepStrictEqual(migrated.map(x => x.id), [ 'one' ])
        Assert.ok(migrated[0].migratedAt instanceof Date)
      })
    })

    describe('#migrate', function () {
      it('should migrate a migration', async function () {
        await da.migrate('two', qi => {
          /* do nothing */
          return Promise.resolve()
        })
        const migrated = await da.getMigratedMigrations()
        Assert.deepStrictEqual(migrated.map(x => x.id), [ 'two' ])
        Assert.ok(migrated[0].migratedAt instanceof Date)
      })
    })

    describe('#rollback', function () {
      it('should rollback a migration', async function () {
        await da.migrate('two', qi => {
          /* do nothing */
          return Promise.resolve()
        })
        await da.rollback('two', qi => {
          /* do nothing */
          return Promise.resolve()
        })
        const migrated = await da.getMigratedMigrations()
        Assert.strictEqual(migrated.length, 0)
      })
    })
  })
})


src/index.ts
import * as Path from 'path'
import { Client } from 'pg'
import * as Flock from '@launchfort/flock'

export class TemplateProvider implements Flock.TemplateProvider {
  readonly migrationTypes = [ 'create-table', 'alter-table', 'other' ]

  provideFileName (migrationType: string) {
    if (this.migrationTypes.indexOf(migrationType) >= 0) {
      return Promise.resolve(Path.join(__dirname, 'templates', migrationType + '.ejs'))
    } else {
      return Promise.reject(Object.assign(
        new Error(`Unsupported migration type [${migrationType}]`),
        { code: 'UNSUPPORTED_MIGRATION_TYPE' }
      ))
    }
  }
}

export class DataAccessProvider implements Flock.DataAccessProvider {
  readonly migrationTableName: string
  readonly acquireLock: boolean
  readonly connectionString: string

  constructor ({ migrationTableName = 'migration', acquireLock = true, connectionString = process.env.DATABASE_URL } = {}) {
    this.migrationTableName = migrationTableName
    this.acquireLock = acquireLock
    this.connectionString = connectionString
  }

  async provide () {
    const lock = 543222
    let locked = false
    const client = new Client({ connectionString: this.connectionString })
    await client.connect()

    if (this.acquireLock) {
      const advisoryLockResult = await client.query({ text: `SELECT pg_try_advisory_lock(${lock})` })
      locked = advisoryLockResult.rows[0].pg_try_advisory_lock === true

      if (!locked) {
        await client.end()
        throw new Error(`Advisory lock "${lock}" could not be acquired.`)
      }
    }

    return new PgDataAccess(client, this.migrationTableName, { lock: locked ? lock : NaN })
  }
}

export class PgDataAccess implements Flock.DataAccess {
  private client: Client
  private qi: PgQueryInterface
  readonly migrationTableName: string
  readonly lock: number

  constructor (client, migrationTableName, { lock = NaN } = {}) {
    this.client = client
    this.qi = new PgQueryInterface(client)
    this.migrationTableName = migrationTableName
    this.lock = lock
  }

  async getMigratedMigrations () {
    const migrationTableExists = await this.qi.tableExists(this.migrationTableName)

    if (!migrationTableExists) {
      return []
    }

    const result = await this.qi.query({
      text: `SELECT id, created_at FROM "${this.migrationTableName}"`
    })
    return (result.rows || []).map(x => {
      return { id: x.id, migratedAt: x.created_at }
    })
  }

  async migrate (migrationId: string, action: (qi: Flock.QueryInterface) => Promise<void>) {
    const migrationTableExists = await this.qi.tableExists(this.migrationTableName)
    const hasMigrated = migrationTableExists ? (await this.hasMigrated(migrationId)) : false

    if (hasMigrated) {
      return
    }

    await this.qi.query({ text: 'BEGIN' })
    try {
      await this.qi.query({
        text:
          `CREATE TABLE IF NOT EXISTS "${this.migrationTableName}" (
            id varchar(512),
            created_at timestamp DEFAULT current_timestamp,
            PRIMARY KEY(id)
          )`
      })
      await action(this.qi)
      await this.qi.query({
        text: `INSERT INTO "${this.migrationTableName}" (id) VALUES($1)`,
        values: [ migrationId ]
      })
      await this.qi.query({ text: 'COMMIT' })
    } catch (error) {
      await this.qi.query({ text: 'ROLLBACK' })
      throw error
    }
  }

  async rollback (migrationId: string, action: (qi: Flock.QueryInterface) => Promise<void>) {
    const migrationTableExists = await this.qi.tableExists(this.migrationTableName)

    if (!migrationTableExists) {
      return
    }

    const hasMigrated = await this.hasMigrated(migrationId)

    if (!hasMigrated) {
      return
    }

    await this.qi.query({ text: 'BEGIN' })
    try {
      await action(this.qi)
      await this.qi.query({
        text: `DELETE FROM "${this.migrationTableName}" WHERE id = $1`,
        values: [ migrationId ]
      })
      await this.qi.query({ text: 'COMMIT' })
    } catch (error) {
      await this.qi.query({ text: 'ROLLBACK' })
      throw error
    }
  }

  async close () {
    if (!isNaN(this.lock) && this.lock !== null) {
      await this.qi.query({ text: `SELECT pg_advisory_unlock(${this.lock})` })
    }
    return this.client.end()
  }

  private async hasMigrated (migrationId: string) {
    const result = await this.qi.query({
      text: `SELECT id FROM "${this.migrationTableName}" WHERE id = $1`,
      values: [ migrationId ]
    })
    return result.rowCount === 1
  }
}

export class PgQueryInterface implements Flock.QueryInterface {
  client: { query(queryObject: { text: string, values?: any[], name?: string }): Promise<Flock.QueryResult> }

  constructor (client: Client) {
    this.client = client
  }

  query (queryObject: { text: string, values?: any[], name?: string }): Promise<Flock.QueryResult> {
    return this.client.query(queryObject)
  }

  async tableExists (tableName: string) {
    // ANSI SQL compliant query. This should work for all RDMS.
    // NOTE: use schema_name() for MSSQL
    const result = await this.query({
      text:
        `SELECT table_name
        FROM   information_schema.tables
        WHERE  table_name = $1
        AND table_schema = current_schema()
        `,
      values: [ tableName ]
    })
    return result.rowCount === 1
  }

  async columnExists (tableName: string, columnName: string) {
    // ANSI SQL compliant query. This should work for all RDMS.
    // NOTE: use schema_name() for MSSQL
    const result = await this.query({
      text:
        `SELECT column_name
        FROM information_schema.columns
        WHERE table_name=$1
        and column_name=$2
        and table_schema = current_schema()`,
      values: [ tableName, columnName ]
    })
    return result.rowCount === 1
  }

  async columnDataType (tableName: string, columnName: string): Promise<string|null> {
    return this.inspectColumn(tableName, columnName).then(col => {
      return col ? col.data_type : null
    })
  }

  private async inspectColumn (tableName: string, columnName: string) {
    // ANSI SQL compliant query. This should work for all RDMS.
    // NOTE: use schema_name() for MSSQL
    const result = await this.query({
      text:
        `SELECT *
        FROM   information_schema.columns
        WHERE  table_name = $1
        AND table_schema = current_schema()
        AND column_name = $2`,
      values: [ tableName, columnName ]
    })
    return result.rowCount === 1 ? result.rows[0] : null
  }
}


lib/index.d.ts
import { Client } from 'pg';
import * as Flock from '@launchfort/flock';
export declare class TemplateProvider implements Flock.TemplateProvider {
    readonly migrationTypes: string[];
    provideFileName(migrationType: string): Promise<string>;
}
export declare class DataAccessProvider implements Flock.DataAccessProvider {
    readonly migrationTableName: string;
    readonly acquireLock: boolean;
    readonly connectionString: string;
    constructor({ migrationTableName, acquireLock, connectionString }?: {
        migrationTableName?: string;
        acquireLock?: boolean;
        connectionString?: string;
    });
    provide(): Promise<PgDataAccess>;
}
export declare class PgDataAccess implements Flock.DataAccess {
    private client;
    private qi;
    readonly migrationTableName: string;
    readonly lock: number;
    constructor(client: any, migrationTableName: any, { lock }?: {
        lock?: number;
    });
    getMigratedMigrations(): Promise<{
        id: any;
        migratedAt: any;
    }[]>;
    migrate(migrationId: string, action: (qi: Flock.QueryInterface) => Promise<void>): Promise<void>;
    rollback(migrationId: string, action: (qi: Flock.QueryInterface) => Promise<void>): Promise<void>;
    close(): Promise<void>;
    private hasMigrated;
}
export declare class PgQueryInterface implements Flock.QueryInterface {
    client: {
        query(queryObject: {
            text: string;
            values?: any[];
            name?: string;
        }): Promise<Flock.QueryResult>;
    };
    constructor(client: Client);
    query(queryObject: {
        text: string;
        values?: any[];
        name?: string;
    }): Promise<Flock.QueryResult>;
    tableExists(tableName: string): Promise<boolean>;
    columnExists(tableName: string, columnName: string): Promise<boolean>;
    columnDataType(tableName: string, columnName: string): Promise<string | null>;
    private inspectColumn;
}


lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Path = require("path");
const pg_1 = require("pg");
class TemplateProvider {
    constructor() {
        this.migrationTypes = ['create-table', 'alter-table', 'other'];
    }
    provideFileName(migrationType) {
        if (this.migrationTypes.indexOf(migrationType) >= 0) {
            return Promise.resolve(Path.join(__dirname, 'templates', migrationType + '.ejs'));
        }
        else {
            return Promise.reject(Object.assign(new Error(`Unsupported migration type [${migrationType}]`), { code: 'UNSUPPORTED_MIGRATION_TYPE' }));
        }
    }
}
exports.TemplateProvider = TemplateProvider;
class DataAccessProvider {
    constructor({ migrationTableName = 'migration', acquireLock = true, connectionString = process.env.DATABASE_URL } = {}) {
        this.migrationTableName = migrationTableName;
        this.acquireLock = acquireLock;
        this.connectionString = connectionString;
    }
    provide() {
        return __awaiter(this, void 0, void 0, function* () {
            const lock = 543222;
            let locked = false;
            const client = new pg_1.Client({ connectionString: this.connectionString });
            yield client.connect();
            if (this.acquireLock) {
                const advisoryLockResult = yield client.query({ text: `SELECT pg_try_advisory_lock(${lock})` });
                locked = advisoryLockResult.rows[0].pg_try_advisory_lock === true;
                if (!locked) {
                    yield client.end();
                    throw new Error(`Advisory lock "${lock}" could not be acquired.`);
                }
            }
            return new PgDataAccess(client, this.migrationTableName, { lock: locked ? lock : NaN });
        });
    }
}
exports.DataAccessProvider = DataAccessProvider;
class PgDataAccess {
    constructor(client, migrationTableName, { lock = NaN } = {}) {
        this.client = client;
        this.qi = new PgQueryInterface(client);
        this.migrationTableName = migrationTableName;
        this.lock = lock;
    }
    getMigratedMigrations() {
        return __awaiter(this, void 0, void 0, function* () {
            const migrationTableExists = yield this.qi.tableExists(this.migrationTableName);
            if (!migrationTableExists) {
                return [];
            }
            const result = yield this.qi.query({
                text: `SELECT id, created_at FROM "${this.migrationTableName}"`
            });
            return (result.rows || []).map(x => {
                return { id: x.id, migratedAt: x.created_at };
            });
        });
    }
    migrate(migrationId, action) {
        return __awaiter(this, void 0, void 0, function* () {
            const migrationTableExists = yield this.qi.tableExists(this.migrationTableName);
            const hasMigrated = migrationTableExists ? (yield this.hasMigrated(migrationId)) : false;
            if (hasMigrated) {
                return;
            }
            yield this.qi.query({ text: 'BEGIN' });
            try {
                yield this.qi.query({
                    text: `CREATE TABLE IF NOT EXISTS "${this.migrationTableName}" (
            id varchar(512),
            created_at timestamp DEFAULT current_timestamp,
            PRIMARY KEY(id)
          )`
                });
                yield action(this.qi);
                yield this.qi.query({
                    text: `INSERT INTO "${this.migrationTableName}" (id) VALUES($1)`,
                    values: [migrationId]
                });
                yield this.qi.query({ text: 'COMMIT' });
            }
            catch (error) {
                yield this.qi.query({ text: 'ROLLBACK' });
                throw error;
            }
        });
    }
    rollback(migrationId, action) {
        return __awaiter(this, void 0, void 0, function* () {
            const migrationTableExists = yield this.qi.tableExists(this.migrationTableName);
            if (!migrationTableExists) {
                return;
            }
            const hasMigrated = yield this.hasMigrated(migrationId);
            if (!hasMigrated) {
                return;
            }
            yield this.qi.query({ text: 'BEGIN' });
            try {
                yield action(this.qi);
                yield this.qi.query({
                    text: `DELETE FROM "${this.migrationTableName}" WHERE id = $1`,
                    values: [migrationId]
                });
                yield this.qi.query({ text: 'COMMIT' });
            }
            catch (error) {
                yield this.qi.query({ text: 'ROLLBACK' });
                throw error;
            }
        });
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!isNaN(this.lock) && this.lock !== null) {
                yield this.qi.query({ text: `SELECT pg_advisory_unlock(${this.lock})` });
            }
            return this.client.end();
        });
    }
    hasMigrated(migrationId) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.qi.query({
                text: `SELECT id FROM "${this.migrationTableName}" WHERE id = $1`,
                values: [migrationId]
            });
            return result.rowCount === 1;
        });
    }
}
exports.PgDataAccess = PgDataAccess;
class PgQueryInterface {
    constructor(client) {
        this.client = client;
    }
    query(queryObject) {
        return this.client.query(queryObject);
    }
    tableExists(tableName) {
        return __awaiter(this, void 0, void 0, function* () {
            // ANSI SQL compliant query. This should work for all RDMS.
            // NOTE: use schema_name() for MSSQL
            const result = yield this.query({
                text: `SELECT table_name
        FROM   information_schema.tables
        WHERE  table_name = $1
        AND table_schema = current_schema()
        `,
                values: [tableName]
            });
            return result.rowCount === 1;
        });
    }
    columnExists(tableName, columnName) {
        return __awaiter(this, void 0, void 0, function* () {
            // ANSI SQL compliant query. This should work for all RDMS.
            // NOTE: use schema_name() for MSSQL
            const result = yield this.query({
                text: `SELECT column_name
        FROM information_schema.columns
        WHERE table_name=$1
        and column_name=$2
        and table_schema = current_schema()`,
                values: [tableName, columnName]
            });
            return result.rowCount === 1;
        });
    }
    columnDataType(tableName, columnName) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.inspectColumn(tableName, columnName).then(col => {
                return col ? col.data_type : null;
            });
        });
    }
    inspectColumn(tableName, columnName) {
        return __awaiter(this, void 0, void 0, function* () {
            // ANSI SQL compliant query. This should work for all RDMS.
            // NOTE: use schema_name() for MSSQL
            const result = yield this.query({
                text: `SELECT *
        FROM   information_schema.columns
        WHERE  table_name = $1
        AND table_schema = current_schema()
        AND column_name = $2`,
                values: [tableName, columnName]
            });
            return result.rowCount === 1 ? result.rows[0] : null;
        });
    }
}
exports.PgQueryInterface = PgQueryInterface;
